File: modules/wem-api/src/main/java/com/enonic/wem/api/content/site/ModuleConfig.java
package com.enonic.wem.api.content.site;


import java.util.Objects;

import com.google.common.base.Preconditions;

import com.enonic.wem.api.data.RootDataSet;
import com.enonic.wem.api.module.ModuleKey;

public final class ModuleConfig
{
    private final ModuleKey module;

    private final RootDataSet config;

    public ModuleConfig( final Builder builder )
    {
        Preconditions.checkNotNull( builder.module, "module cannot be null" );
        Preconditions.checkNotNull( builder.config, "config cannot be null" );
        this.module = builder.module;
        this.config = builder.config;
    }

    public ModuleKey getModule()
    {
        return module;
    }

    public RootDataSet getConfig()
    {
        return config;
    }

    @Override
    public boolean equals( final Object o )
    {
        if ( this == o )
        {
            return true;
        }
        if ( o == null || getClass() != o.getClass() )
        {
            return false;
        }

        final ModuleConfig that = (ModuleConfig) o;

        return Objects.equals( this.module, that.module ) && Objects.equals( this.config, that.config );
    }

    @Override
    public int hashCode()
    {
        return Objects.hash( module, config );
    }

    public static Builder newModuleConfig()
    {
        return new Builder();
    }

    public static class Builder
    {
        private ModuleKey module;

        private RootDataSet config;

        public Builder module( ModuleKey value )
        {
            this.module = value;
            return this;
        }

        public Builder config( RootDataSet value )
        {
            this.config = value;
            return this;
        }

        public ModuleConfig build()
        {
            return new ModuleConfig( this );
        }
    }
}

File: modules/wem-admin-ui/src/main/resources/web/admin/common/js/content/site/ModuleConfig.ts
module api.content.site {

    export class ModuleConfig implements api.Equitable, api.Cloneable {

        private moduleKey: api.module.ModuleKey;

        private config: api.data.RootDataSet;

        constructor(builder: ModuleConfigBuilder) {
            this.moduleKey = builder.moduleKey;
            this.config = builder.config;
        }

        getModuleKey(): api.module.ModuleKey {
            return this.moduleKey;
        }

        getConfig(): api.data.RootDataSet {
            return this.config;
        }

        setConfig(config: api.data.RootDataSet) {
            this.config = config;
        }

        toJson(): Object {
            return {
                moduleKey: this.moduleKey.toString(),
                config: this.config.toJson()
            }
        }

        toData(): api.data.RootDataSet {
            var data = new api.data.RootDataSet();
            data.addProperty("moduleKey", new api.data.Value(this.moduleKey.getName(), api.data.type.ValueTypes.STRING));
            data.addProperty("config", new api.data.Value(this.config, api.data.type.ValueTypes.DATA));
            return data;
        }

        equals(o: api.Equitable): boolean {

            if (!api.ObjectHelper.iFrameSafeInstanceOf(o, ModuleConfig)) {
                return false;
            }

            var other = <ModuleConfig>o;

            if (!api.ObjectHelper.equals(this.moduleKey, other.moduleKey)) {
                return false;
            }

            if (!api.ObjectHelper.equals(this.config, other.config)) {
                return false;
            }

            return true;
        }

        clone(): ModuleConfig {

            return new ModuleConfigBuilder(this).build();
        }
    }

    export class ModuleConfigBuilder {

        moduleKey: api.module.ModuleKey;

        config: api.data.RootDataSet;

        constructor(source?: ModuleConfig) {
            if (source) {
                this.moduleKey = source.getModuleKey();
                this.config = source.getConfig() ? source.getConfig().clone() : null;
            }
        }

        fromData(data: api.data.RootDataSet): ModuleConfigBuilder {
            api.util.assertNotNull(data, "data cannot be null");
            var moduleKey = api.module.ModuleKey.fromString(data.getProperty("moduleKey").getString());
            var moduleConfig = data.getProperty("config").getData();
            this.setModuleKey(moduleKey);
            this.setConfig(moduleConfig);
            return this;
        }

        setModuleKey(value: api.module.ModuleKey): ModuleConfigBuilder {
            this.moduleKey = value;
            return this;
        }

        setConfig(value: api.data.RootDataSet): ModuleConfigBuilder {
            this.config = value;
            return this;
        }

        build(): ModuleConfig {
            return new ModuleConfig(this);
        }
    }

}